### PR TITLE
Genotyper work 

### PR DESCRIPTION
**Pull request by akiezun**
first cut of the genotyper work. 
@davidbenjamin I need your help in migrating log10 to ln. I had to revert this on change on the branch. Please take over the branch and migrate back. I left the revert commit for you to see what i did.

### CODE REVIEW

**Comment by akiezun** [comment]
@davidbenjamin any progress here? i'm blocked on this issue for the haplotype caller

**Comment by davidbenjamin** [comment]
@akiezun I have switched a lot so far; currently I'm stuck on `GenotypeLikelihoodCalculator`, which relies on `GenotypeLikelihoods` in htsjdk, which is log10.  I could write a simple wrapper class to present a natural log interface.  Is there a better solution?

**Comment by akiezun** [comment]
yes, please try wrapping. We can't push that down to htsjdk because it would be too disruptive for all clients of that library

**Comment by davidbenjamin** [comment]
@akiezun All tests are passing.  I handle the log10s from htsjdk with a couple of atrocious hacks, though, and I'll solicit @lbergelson's advice on writing a better facade before handing it back to you.

**Comment by davidbenjamin** [comment]
In discussion with @akiezun, @ldgauthier, and @lbergelson, we decided that the natural log crusade isn't worth it.  When this branch is merged it will roll back gatk to log10.

**Comment by davidbenjamin** [comment]
It was fun while it lasted.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/CustomAFPriorProvider.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Custom priors provided as an array.
+ */
+public final class CustomAFPriorProvider extends AFPriorProvider {
+
+    private final double[] priors;
+
+    /**
+     *
+     * @param priorValues must exactly the number of genomes in the samples (the total ploidy).
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/CustomAFPriorProvider.java]: javadoc missing "equal"

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+     * calculation request calling {@link #ensureReadCapacity(int) ensureReadCapacity}.<p/>
+     */
+    private int readCapacity = -1;
+
+    /**
+     * Caches the log10 of the first few integers up to the ploidy supported by the calculator.
+     * <p>This is in fact a shallow copy if {@link GenotypeLikelihoodCalculators#ploidyLog10}</p> and is not meant to be modified by
+     * this class. </p>
+     */
+    private final double[] log10;
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java]: `MathUtils` already has a LogCache.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+     */
+    private double[] readGenotypeLikelihoodComponents;
+
+    /**
+     * Creates a new calculator providing its ploidy and number of genotyping alleles.
+     */
+    protected GenotypeLikelihoodCalculator(final int ploidy, final int alleleCount,
+                                           final int[][] alleleFirstGenotypeOffsetByPloidy,
+                                           final GenotypeAlleleCounts[][] genotypeTableByPloidy,
+                                           final double[] ploidyLog10) {
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java]: Same comment about the `LogCache`.  There's code in several places where a `ploidyLog10` array is calculated in order to be passed to this method, so make sure to remove those calculations.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java [hunk 4]
```diff
@@ -0,0 +1,620 @@
...
+     *   genotype (0-based).
+     */
+    private double[] genotypeLikelihoods(final double[][] readLikelihoodsByGenotypeIndex, final int readCount) {
+        final double[] result = new double[genotypeCount];
+        final double denominator = readCount * log10[ploidy]; // instead of dividing each read likelihood by ploidy
+         // ( so subtract log10(ploidy) )  we multiply them all and the divide by ploidy^readCount (so substract readCount * log10(ploidy) )
+        for (int g = 0; g < genotypeCount; g++) {
+            final double[] likelihoodsByRead = readLikelihoodsByGenotypeIndex[g];
+            double s = - denominator;
+            for (int r = 0; r < readCount; r++) {
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java]: This should be written using an array sum method in `MathUtils`.  If no method exists for summing the first `readCount` values of `likelihoodsByRead` it should be written.  This code would then become `result[g] = MathUtils.sum(likelihoodsByRead[g], readCount) - denominator`, which is much clearer.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+     *     Its initial value indicates the initial capacity of the shared {@link #genotypeTableByPloidy}. Feel free
+     *     to change it to anything reasonable that is non-negative.
+     * </p>
+     */
+    private int maximumPloidy = 2; // its initial value is the initial capacity of the shared tables.
+
+    /**
+     * Cached log10 values for the first integer up to the maximum ploidy requested thus far.
+     */
+    private double[] ploidyLog10;
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java]: Same as above: use `MathUtils.LogCache`.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+
+    // Initialize {@link #ploidyLog10}.
+    {
+        ploidyLog10 = new double[maximumPloidy + 1];
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java]: No need to reinvent the `LogCache` wheel.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+        }
+
+        final int newMaximumPloidy = Math.max(maximumPloidy, requestedMaximumPloidy);
+        final int newMaximumAllele = Math.max(maximumAllele, requestedMaximumAllele);
+
+        // Update tables first.
+        alleleFirstGenotypeOffsetByPloidy = buildAlleleFirstGenotypeOffsetTable(newMaximumPloidy,newMaximumAllele);
+        genotypeTableByPloidy = buildGenotypeAlleleCountsTable(newMaximumPloidy,newMaximumAllele,alleleFirstGenotypeOffsetByPloidy);
+
+        if (needsToExpandPloidyCapacity) {
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java]: This code will go away if you use `MathUtils.LogCache`.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+        if (needsToExpandAlleleCapacity) {
+            maximumAllele = requestedMaximumAllele;
+        }
+        if (needsToExpandPloidyCapacity) {
+            maximumPloidy = requestedMaximumPloidy;
+        }
+    }
+
+    /**
+     * Extends the existing {@link #ploidyLog10} with more log10 as needed by maximum-ploidy expansion.
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java]: Look how much code can be deleted by using `MathUtils.LogCache`!

**Comment by akiezun** [comment]
@davidbenjamin is this back with me?

**Comment by davidbenjamin** [comment]
@akiezun Yikes, forgot to reassign!  Yes, it's yours again.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/CustomAFPriorProvider.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/CustomAFPriorProvider.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java]: done

**Comment by akiezun** [comment]
ugh, too much pain to bring it back to life without a rebase (dataflow code is now dead and many jars mentioned in this branch are now gone). Will rebase first. It's going to be painful too but necessary in the long run anyway

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java (hunk 8 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculators.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java (hunk 4 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculator.java]: done

**Comment by akiezun** [comment]
@vruano can you review? This is just the first cut of this work and whole bunch of code was left close to as-is.

**Comment by vruano** [comment]
Done with review back to @akiezun

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/ReferenceDiploidExactAFCalculator.java [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            final ExactACset set = new ExactACset(numChr/2 +1, index);
+            indexesToACset.put(index, set);
+            ACqueue.add(set);
+        }
+
+        // push data from the dependency to the new set
+        pushData(indexesToACset.get(index), dependentSet, PLsetIndex, genotypeLikelihoods);
+    }
+
+    private static void pushData(final ExactACset targetSet,
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/ReferenceDiploidExactAFCalculator.java]: note to self: port improvements from https://github.com/broadinstitute/gsa-unstable/pull/1270

**Pull request merged**